<h1> Комментарий тимлида </h1>

👋 **Привет, Татьяна!** 

✨ Поздравляю с началом работы над финальным проектом!

Меня зовут Арсен Абдулин и я буду твоим ревьювером. Предлагаю общаться на «ты», если не против =) 

При проверке работ я делаю следующие комментарии:

<div class="alert alert-success">
<b>✔️ Зеленым цветом</b> отмечены удачные решения.</div>

<div class="alert alert-warning">
<b>⚠️ Желтым цветом</b> я отметил пункты, которые в следующий раз можно сделать по-другому. Одно-два таких замечания в проекте допускается, но если их много — проект следует доработать. </div>

<div class="alert alert-danger">
<b>🚫 Красным цветом</b> отмечены критические замечания, которые необходимо поправить, чтобы принять проект. </div>

Если какие-то моменты в задании для тебя были непонятны и у тебя есть ко мне вопросы — смело спрашивай 😊 Также ты можешь доработать места, где есть желтые комментарии в проекте (однако, это не обязательно).

Предлагаю работать в диалоге: если ты решишь что-то поменять по моим рекомендациям — пиши об этом (выбери для своих комментариев определенный цвет - так мне будет легче увидеть изменения). Пожалуйста не перемещай, не изменяй и не удаляй мои комментарии. Все это поможет сделать проверку твоего проекта оперативнее.

<div class="alert alert-info"> <b>ℹ️ Комментарий студента: </b> Пример комментария. </div>

# План проекта

<div class="alert alert-success">
<b>✔️ Комментарий тимлида:</b> Хорошо, когда есть подробное описание проекта и поставлена цель — так постороннему человеку будет проще ознакомиться!
   
</div>

# Проект домена Телеком
# «Прогнозирование оттока клиентов»

Описание проекта

Оператор связи «Ниединогоразрыва.ком» хочет научиться прогнозировать отток клиентов. Если выяснится, что пользователь планирует уйти, ему будут предложены промокоды и специальные условия. 

**Цель проекта** - прогнозирование оттока клиентов.   

**Наша задача** - создать прототип модели машинного обучения, который позволит оператору связи своевременно выявлять клиентов, планирующих уйти в ближайшее время. Это позволит компании предложить клиентам промокоды и специальные условия, тем самым сохранить клиента и увеличить прибыль.  

Команда оператора собрала персональные данные о некоторых клиентах, информацию об их тарифах и договорах.

**Описание услуг**

Оператор предоставляет два основных типа услуг: 

1. Стационарную телефонную связь. Возможно подключение телефонного аппарата к нескольким линиям одновременно.
2. Интернет. Подключение может быть двух типов: через телефонную линию (DSL*,* от англ. *digital subscriber line*, «цифровая абонентская линия») или оптоволоконный кабель (*Fiber optic*).  

Также доступны такие услуги:

- Интернет-безопасность: антивирус (*DeviceProtection*) и блокировка небезопасных сайтов (*OnlineSecurity*);
- Выделенная линия технической поддержки (*TechSupport*);
- Облачное хранилище файлов для резервного копирования данных (*OnlineBackup*);
- Стриминговое телевидение (*StreamingTV*) и каталог фильмов (*StreamingMovies*).

За услуги клиенты могут платить каждый месяц или заключить договор на 1–2 года. Доступны различные способы расчёта и возможность получения электронного чека.

Данные состоят из файлов, полученных из разных источников:
* contract_new.csv — информация о договоре;
* personal_new.csv — персональные данные клиента;
* internet_new.csv — информация об интернет-услугах;
* phone_new.csv — информация об услугах телефонии. 

**Описание данных, предоставленное заказчиком:**

BeginDate – дата начала пользования услугами,   
EndDate – дата окончания пользования услугами,   
Type – тип оплаты: ежемесячный, годовой и тд,   
PaperlessBilling – безналичный расчет,   
PaymentMethod – способ оплаты,   
MonthlyCharges – ежемесячные траты на услуги,   
TotalCharges – всего потрачено денег на услуги   
Dependents – наличие иждивенцев   
Senior Citizen – наличие пенсионного статуса по возрасту   
Partner – наличие супруга(и)   
MultipleLines – наличие возможности ведения параллельных линий во время   
звонка   

**План проекта:**

**1. Загрузка и предобработка данных:**

* изучим общую информацию о полученных датасетах, посмотрим типы данных, пропуски, дубликаты;
* объединим все датасеты в один для удобства работы и последующего анализа;
* если среди признаков есть пропуски,то заполним их и обработаем аномалии;
* сгенерируем новый признак;
* после генерации новых признаков для предотвращения утечки данных удалим признаки BeginDate и EndDate; 
* краткий вывод

**2. Исследовательский анализ данных (EDA):**

* изучим целевой признак - проверим дисбаланс;
* проверим корреляцию числовых признаков на наличие мультиколлинеарности;
* рассмотрим влияние категориальных признаков на таргет - сравним 2 группы клиентов, тех кто ушел и лояльных (размер ежемесячных платежей, наличие супруга, иждивенцев, количество дополнительных услуг ит.д.) на факт ухода клиента;
* для наглядности построим графики зависимостей;
* удалим имеющиеся дубликаты;
* краткий вывод по EDA.

**3. Подготовка данных для обучения моделей:**

* введем константу RANDOM STATE = 220523;
* создадим переменные для признаков и целевого признака;
* разделим исходные данные на обучающую и тестовую выборки (test_size = 0.25);
* кодирование данных для линейных моделей;
* масштабирование данных для линейных моделей.

**4. Обучение моделей:**  

* обучим разные модели (линейные и бустинг)
* для оптимального подбора гиперпараметров будем использовать GridSearchCV;
* для оценки качества моделей будем использовать метрику AUC-ROC,ее значение должно быть >= 0.85; 
  в качестве дополнительной метрики будем использовать accuracy.   
* выбор лучшей модели;
* тестирование лучшей модели на тестовой выборке;
* анализ результатов, подсчет дополнительных метрик при необходимости, проверка на адекватность;
* построение ROC-кривой.

**5. Общий вывод.**

<div class="alert alert-success">
<b>✔️ Комментарий тимлида:</b> План работ последовательный и понятный, все хорошо!
    
</div>

**Вопросы и комментарии к тимлиду:**
    
Арсен, на данном этапе мне все понятно, вопросов нет.


<div class="alert alert-success">
<b>✔️ Комментарий тимлида:</b> Хорошо =)
    
</div>

## Загрузка и предобработка данных

### Для предварительного анализа данных загрузим имеющиеся датасеты и рассмотрим их особенности.

In [1]:
# предварительно импортируем только библиотеку pandas и pandas-profiling

import pandas as pd
!pip install pandas-profiling
from pandas_profiling import ProfileReport



Ссылки, где хранятся датасеты

https://code.s3.yandex.net/datasets/contract_new.csv   
https://code.s3.yandex.net/datasets/personal_new.csv   
https://code.s3.yandex.net/datasets/internet_new.csv   
https://code.s3.yandex.net/datasets/phone_new.csv      
    

In [2]:
try:
    contract = pd.read_csv('/datasets/contract_new.csv')
    personal = pd.read_csv('/datasets/personal_new.csv')
    internet = pd.read_csv('/datasets/internet_new.csv')
    phone = pd.read_csv('/datasets/phone_new.csv')
except:
    contract = pd.read_csv('https://code.s3.yandex.net/datasets/contract_new.csv')
    personal = pd.read_csv('https://code.s3.yandex.net/datasets/personal_new.csv')
    internet = pd.read_csv('https://code.s3.yandex.net/datasets/internet_new.csv')
    phone = pd.read_csv('https://code.s3.yandex.net/datasets/phone_new.csv')

In [4]:
# напишем функция для создания отчета о датасетах с помощью ProfileReport
def view_information(data):
    ProfileReport(data, title="Data Profiling Report")
    print(profile.to_widgets())

In [5]:
view_information(contract)

NameError: name 'profile' is not defined

In [ ]:
view_information(personal)

In [ ]:
view_information(internet)

In [ ]:
view_information(phone)

Вывод по загрузке данных

В результате загрузки данных, полученных от телеком компании, мы получили 4 датасета с разным количеством строк.
Основным является датасет с информацией по контрактам, вторым по значимости будет второй датасет с персональными данными клиентов. У этих датасетов количество строк совпадает и составляет 7043.

Также у нас есть данные какими интернет услугами пользуется клиенты и какими услугами телефонии. Здесь количество строк отличается, это связано с тем, что  кто-то из клиентов пользуется только интернетом, а кто-то только услугами телефонии.

Соответственно, для полноценного анализа данных мы объединим все 4 датасета по `customerID`.

Также мы видим, что в столбце `EndDate` не везде отмечена дата окончания пользования услугами. Это говорит, о том, что у нас есть сведения не только об ушедших,но еще и о лояльных клиентах.

В некоторых столбцах (например, `TotalCharges`) требуется поменять тип данных.

Перед нами стоит задача бинарной классификации. Далее приступим к более детальной предобработке данных, в соответствии с нашим планом.

<div class="alert alert-success">
<b>✔️ Комментарий тимлида:</b> Предварительный анализ данных выполнен успешно!
    
Видно, что нужна будет предобработка данных
    
</div>

<div class="alert alert-success">
<b>✔️ Комментарий тимлида:</b> План работ принят, можно приступать к основному проекту.
    
Сохрани пожалуйста наши с тобой комментарии, а проект продолжай в этом же файле. Успехов!
</div>

<div class="alert alert-info"> <b>ℹ️ Комментарий студента: </b>Привет, Арсен, спасибо за обратную связь. Продолжаю работу). </div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Всегда пожалуйста 😊
    
</div>

<h1> Комментарий ревьювера </h1>

<div class="alert alert-success">
    
👋 Татьяна, привет еще раз! Это снова Арсен Абдулин — и я буду делать ревью основного проекта.
    
Давай посмотрим, что получилось.
    
</div>

### Объединим все датасеты в один для удобства работы и последующего анализа

Импортируем необходимые для анализа библиотеки и структуры данных 

In [6]:
!pip install -q phik
!pip install -U -q scikit-learn
#!pip install -U -q matplotlib
#!pip install -U -q seaborn

In [7]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
import time
import catboost as cb
import phik

from phik.report import plot_correlation_matrix
from phik import report
from pandas import DataFrame

from sklearn.preprocessing import StandardScaler
from sklearn.compose import make_column_transformer

from sklearn import tree
from sklearn.model_selection import (
    train_test_split,
    GridSearchCV,
    KFold,
    RandomizedSearchCV
)

from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import Ridge
from sklearn.dummy import DummyClassifier
from catboost import CatBoostClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score, 
    recall_score, 
    f1_score, 
    roc_auc_score,
    roc_curve, 
    precision_recall_curve,
    confusion_matrix,
    ConfusionMatrixDisplay
)

from sklearn.preprocessing import (
    OneHotEncoder,
    OrdinalEncoder,
    StandardScaler
)


from sklearn.model_selection import (
    cross_validate,
    cross_val_score
)

from sklearn.pipeline import Pipeline
from sklearn.pipeline import make_pipeline
warnings.filterwarnings("ignore")

RANDOM_STATE = 220523

Объединим датасеты по общему столбцу `customerID`.

In [8]:
data1 = pd.merge(contract, personal,how ='left', on=('customerID'))

In [9]:
data2 = pd.merge(data1,internet,how ='left', on=('customerID'))

In [10]:
data = pd.merge(data2,phone,how ='left', on=('customerID'))

In [11]:
# проверим размер итоговой таблицы
data.shape

(7043, 20)

In [12]:
data.head()

customerID   BeginDate EndDate            Type PaperlessBilling  \
0  7590-VHVEG  2020-01-01      No  Month-to-month              Yes   
1  5575-GNVDE  2017-04-01      No        One year               No   
2  3668-QPYBK  2019-10-01      No  Month-to-month              Yes   
3  7795-CFOCW  2016-05-01      No        One year               No   
4  9237-HQITU  2019-09-01      No  Month-to-month              Yes   

               PaymentMethod  MonthlyCharges TotalCharges  gender  \
0           Electronic check           29.85        31.04  Female   
1               Mailed check           56.95      2071.84    Male   
2               Mailed check           53.85       226.17    Male   
3  Bank transfer (automatic)           42.30       1960.6    Male   
4           Electronic check           70.70        353.5  Female   

   SeniorCitizen Partner Dependents InternetService OnlineSecurity  \
0              0     Yes         No             DSL             No   
1              0      No         No             DSL            Yes   
2              0      No         No             DSL            Yes   
3              0      No         No             DSL            Yes   
4              0      No         No     Fiber optic             No   

  OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
0          Yes               No          No          No              No   
1           No              Yes          No          No              No   
2          Yes               No          No          No              No   
3           No              Yes         Yes          No              No   
4           No               No          No          No              No   

  MultipleLines  
0           NaN  
1            No  
2            No  
3           NaN  
4            No

Перед нами стоит задача бинарной классификации. Явного таргета у нас нет, поэтому нам следует сгенерировать новый бинарный признак по столбцу `EndDate`. Ранее мы выяснили, что в данном столбце есть конкретные даты окончания пользования услугами и стоит отметка 'No'. Следовательно создадим новый столбец, где '1' - клиент ушел, а '0' - клиент действующий.


<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Отлично, получили целевую переменную!
    
</div>

In [13]:
data['churn'] = np.where(data['EndDate']== 'No',0,1)

In [14]:
data.head(15)

customerID   BeginDate     EndDate            Type PaperlessBilling  \
0   7590-VHVEG  2020-01-01          No  Month-to-month              Yes   
1   5575-GNVDE  2017-04-01          No        One year               No   
2   3668-QPYBK  2019-10-01          No  Month-to-month              Yes   
3   7795-CFOCW  2016-05-01          No        One year               No   
4   9237-HQITU  2019-09-01          No  Month-to-month              Yes   
5   9305-CDSKC  2019-03-01          No  Month-to-month              Yes   
6   1452-KIOVK  2018-04-01          No  Month-to-month              Yes   
7   6713-OKOMC  2019-04-01          No  Month-to-month               No   
8   7892-POOKP  2017-07-01          No  Month-to-month              Yes   
9   6388-TABGU  2014-12-01  2017-05-01        One year               No   
10  9763-GRSKD  2019-01-01          No  Month-to-month              Yes   
11  7469-LKBCI  2018-10-01          No        Two year               No   
12  8091-TTVAX  2015-04-01          No        One year               No   
13  0280-XJGEX  2015-09-01          No  Month-to-month              Yes   
14  5129-JLPIS  2018-01-01          No  Month-to-month              Yes   

                PaymentMethod  MonthlyCharges TotalCharges  gender  \
0            Electronic check           29.85        31.04  Female   
1                Mailed check           56.95      2071.84    Male   
2                Mailed check           53.85       226.17    Male   
3   Bank transfer (automatic)           42.30       1960.6    Male   
4            Electronic check           70.70        353.5  Female   
5            Electronic check           99.65      1150.96  Female   
6     Credit card (automatic)           89.10      2058.21    Male   
7                Mailed check           29.75       300.48  Female   
8            Electronic check          104.80      3573.68  Female   
9   Bank transfer (automatic)           56.15      1628.35    Male   
10               Mailed check           49.95       649.35    Male   
11    Credit card (automatic)           18.95        312.3    Male   
12    Credit card (automatic)          100.35      6111.31    Male   
13  Bank transfer (automatic)          103.70       5496.1    Male   
14           Electronic check          105.50       2637.5    Male   

    SeniorCitizen  ... Dependents InternetService OnlineSecurity OnlineBackup  \
0               0  ...         No             DSL             No          Yes   
1               0  ...         No             DSL            Yes           No   
2               0  ...         No             DSL            Yes          Yes   
3               0  ...         No             DSL            Yes           No   
4               0  ...         No     Fiber optic             No           No   
5               0  ...         No     Fiber optic             No           No   
6               0  ...        Yes     Fiber optic             No          Yes   
7               0  ...         No             DSL            Yes           No   
8               0  ...         No     Fiber optic             No           No   
9               0  ...        Yes             DSL            Yes          Yes   
10              0  ...        Yes             DSL            Yes           No   
11              0  ...         No             NaN            NaN          NaN   
12              0  ...         No     Fiber optic             No           No   
13              0  ...         No     Fiber optic             No          Yes   
14              0  ...         No     Fiber optic            Yes           No   

   DeviceProtection TechSupport StreamingTV StreamingMovies MultipleLines  \
0                No          No          No              No           NaN   
1               Yes          No          No              No            No   
2                No          No          No              No            No   
3               Yes         Yes          No              No           NaN   


In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
 8   gender            7043 non-null   object 
 9   SeniorCitizen     7043 non-null   int64  
 10  Partner           7043 non-null   object 
 11  Dependents        7043 non-null   object 
 12  InternetService   5517 non-null   object 
 13  OnlineSecurity    5517 non-null   object 
 14  OnlineBackup      5517 non-null   object 
 15  DeviceProtection  5517 non-null   object 
 16  TechSupport       5517 non-null   object 


In [16]:
view_information(data)

NameError: name 'profile' is not defined

После объединения датасетов и создания столбца с таргетом, мы получили 21 столбец с 7043 строками. 
Мы видим наличие технологических пропусков в столбцах `InternetService`, `OnlineSecurity`,`OnlineBackup`,`DeviceProtection`,`TechSupport`, `StreamingTV`,`StreamingMovies`,`MultipleLines` связанных с тем, что часть клиентов пользуется интернетом, а другие телефонией. 
Поэтому пропуски в данных столбцах заменим "заглушкой" - 'No'.


<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Как рекомендация, пропуски можно заполнить значением No, так будет меньше лишних категорий
    
</div>

<div class="alert alert-info"> <b>ℹ️ Комментарий студента: </b> Привет, Арсен! Скажи пожалуйста, то есть если мы заполним как No, то есть вероятность что может улучшиться метрика?</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера v2:</b> Как минимум после кодирования будет меньше признаков и модель будет обучаться быстрее
    
В случае бинарных признаков, категория not_used будет дополнительной, а в случае No останутся только две категории, Yes/No
    
Для признака InternetService будут значения DSL/FiberOptic/No
    
</div>

<div class="alert alert-info"> <b>ℹ️ Комментарий студента2: </b> Привет, Арсен!  Поняла, спасибо! </div>

<div class="alert alert-success">
<b>✔️ Комментарий тимлида v2:</b> Хорошо =)
    
</div>

In [17]:
data = data.fillna('No')

In [18]:
data.head(15)

customerID   BeginDate     EndDate            Type PaperlessBilling  \
0   7590-VHVEG  2020-01-01          No  Month-to-month              Yes   
1   5575-GNVDE  2017-04-01          No        One year               No   
2   3668-QPYBK  2019-10-01          No  Month-to-month              Yes   
3   7795-CFOCW  2016-05-01          No        One year               No   
4   9237-HQITU  2019-09-01          No  Month-to-month              Yes   
5   9305-CDSKC  2019-03-01          No  Month-to-month              Yes   
6   1452-KIOVK  2018-04-01          No  Month-to-month              Yes   
7   6713-OKOMC  2019-04-01          No  Month-to-month               No   
8   7892-POOKP  2017-07-01          No  Month-to-month              Yes   
9   6388-TABGU  2014-12-01  2017-05-01        One year               No   
10  9763-GRSKD  2019-01-01          No  Month-to-month              Yes   
11  7469-LKBCI  2018-10-01          No        Two year               No   
12  8091-TTVAX  2015-04-01          No        One year               No   
13  0280-XJGEX  2015-09-01          No  Month-to-month              Yes   
14  5129-JLPIS  2018-01-01          No  Month-to-month              Yes   

                PaymentMethod  MonthlyCharges TotalCharges  gender  \
0            Electronic check           29.85        31.04  Female   
1                Mailed check           56.95      2071.84    Male   
2                Mailed check           53.85       226.17    Male   
3   Bank transfer (automatic)           42.30       1960.6    Male   
4            Electronic check           70.70        353.5  Female   
5            Electronic check           99.65      1150.96  Female   
6     Credit card (automatic)           89.10      2058.21    Male   
7                Mailed check           29.75       300.48  Female   
8            Electronic check          104.80      3573.68  Female   
9   Bank transfer (automatic)           56.15      1628.35    Male   
10               Mailed check           49.95       649.35    Male   
11    Credit card (automatic)           18.95        312.3    Male   
12    Credit card (automatic)          100.35      6111.31    Male   
13  Bank transfer (automatic)          103.70       5496.1    Male   
14           Electronic check          105.50       2637.5    Male   

    SeniorCitizen  ... Dependents InternetService OnlineSecurity OnlineBackup  \
0               0  ...         No             DSL             No          Yes   
1               0  ...         No             DSL            Yes           No   
2               0  ...         No             DSL            Yes          Yes   
3               0  ...         No             DSL            Yes           No   
4               0  ...         No     Fiber optic             No           No   
5               0  ...         No     Fiber optic             No           No   
6               0  ...        Yes     Fiber optic             No          Yes   
7               0  ...         No             DSL            Yes           No   
8               0  ...         No     Fiber optic             No           No   
9               0  ...        Yes             DSL            Yes          Yes   
10              0  ...        Yes             DSL            Yes           No   
11              0  ...         No              No             No           No   
12              0  ...         No     Fiber optic             No           No   
13              0  ...         No     Fiber optic             No          Yes   
14              0  ...         No     Fiber optic            Yes           No   

   DeviceProtection TechSupport StreamingTV StreamingMovies MultipleLines  \
0                No          No          No              No            No   
1               Yes          No          No              No            No   
2                No          No          No              No            No   
3               Yes         Yes          No              No            No   


In [19]:
# проверим наличие пропусков после замены
data.isna().sum() 

customerID          0
BeginDate           0
EndDate             0
Type                0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
MultipleLines       0
churn               0
dtype: int64

In [20]:
# проверим дубликаты по customerID
data['customerID'].duplicated().sum()

0

In [21]:
data.duplicated().sum()

0

У нас есть данные о начале и окончании договора, соответственно мы можем создать еще один признак для нашей модели - 'срок пользования'. Для этого мы заменим в столбце `EndDate` строки со значением 'No' на дату выгрузки '2020-02-01'. 

In [22]:
# заполним 'No' датой на которую предоставлены данные 2020-02-01
data.loc[(data['EndDate'] == 'No'),'EndDate'] = '2020-02-01'

In [23]:
data.head()

customerID   BeginDate     EndDate            Type PaperlessBilling  \
0  7590-VHVEG  2020-01-01  2020-02-01  Month-to-month              Yes   
1  5575-GNVDE  2017-04-01  2020-02-01        One year               No   
2  3668-QPYBK  2019-10-01  2020-02-01  Month-to-month              Yes   
3  7795-CFOCW  2016-05-01  2020-02-01        One year               No   
4  9237-HQITU  2019-09-01  2020-02-01  Month-to-month              Yes   

               PaymentMethod  MonthlyCharges TotalCharges  gender  \
0           Electronic check           29.85        31.04  Female   
1               Mailed check           56.95      2071.84    Male   
2               Mailed check           53.85       226.17    Male   
3  Bank transfer (automatic)           42.30       1960.6    Male   
4           Electronic check           70.70        353.5  Female   

   SeniorCitizen  ... Dependents InternetService OnlineSecurity OnlineBackup  \
0              0  ...         No             DSL             No          Yes   
1              0  ...         No             DSL            Yes           No   
2              0  ...         No             DSL            Yes          Yes   
3              0  ...         No             DSL            Yes           No   
4              0  ...         No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies MultipleLines churn  
0               No          No          No              No            No     0  
1              Yes          No          No              No            No     0  
2               No          No          No              No            No     0  
3              Yes         Yes          No              No            No     0  
4               No          No          No              No            No     0  

[5 rows x 21 columns]

Далее создадим новый признак - срок пользования услугой в днях `term`. Для этого еще изменим тип данных в столбцах на 'datetime64'.

In [24]:
# приведем к типу "дата"
data['BeginDate'] = data['BeginDate'].astype('datetime64')
data['EndDate'] = data['EndDate'].astype('datetime64')

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Хорошо, так мы получим новый полезный признак для обучения
    
</div>

In [25]:
# рассчитаем срок пользования в днях
data['term'] = (data['EndDate']-data['BeginDate']).dt.days

In [26]:
data.head()

customerID  BeginDate    EndDate            Type PaperlessBilling  \
0  7590-VHVEG 2020-01-01 2020-02-01  Month-to-month              Yes   
1  5575-GNVDE 2017-04-01 2020-02-01        One year               No   
2  3668-QPYBK 2019-10-01 2020-02-01  Month-to-month              Yes   
3  7795-CFOCW 2016-05-01 2020-02-01        One year               No   
4  9237-HQITU 2019-09-01 2020-02-01  Month-to-month              Yes   

               PaymentMethod  MonthlyCharges TotalCharges  gender  \
0           Electronic check           29.85        31.04  Female   
1               Mailed check           56.95      2071.84    Male   
2               Mailed check           53.85       226.17    Male   
3  Bank transfer (automatic)           42.30       1960.6    Male   
4           Electronic check           70.70        353.5  Female   

   SeniorCitizen  ... InternetService OnlineSecurity OnlineBackup  \
0              0  ...             DSL             No          Yes   
1              0  ...             DSL            Yes           No   
2              0  ...             DSL            Yes          Yes   
3              0  ...             DSL            Yes           No   
4              0  ...     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies MultipleLines  \
0               No          No          No              No            No   
1              Yes          No          No              No            No   
2               No          No          No              No            No   
3              Yes         Yes          No              No            No   
4               No          No          No              No            No   

  churn  term  
0     0    31  
1     0  1036  
2     0   123  
3     0  1371  
4     0   153  

[5 rows x 22 columns]

In [27]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   customerID        7043 non-null   object        
 1   BeginDate         7043 non-null   datetime64[ns]
 2   EndDate           7043 non-null   datetime64[ns]
 3   Type              7043 non-null   object        
 4   PaperlessBilling  7043 non-null   object        
 5   PaymentMethod     7043 non-null   object        
 6   MonthlyCharges    7043 non-null   float64       
 7   TotalCharges      7043 non-null   object        
 8   gender            7043 non-null   object        
 9   SeniorCitizen     7043 non-null   int64         
 10  Partner           7043 non-null   object        
 11  Dependents        7043 non-null   object        
 12  InternetService   7043 non-null   object        
 13  OnlineSecurity    7043 non-null   object        
 14  OnlineBackup      7043 n

Далее посмотрим более детально на типы данных в столбцах. Требуется изменить тип данных в столбце `TotalCharges`, т.к. сумма затрат не может быть типом object.

In [28]:
# применим метод to_numeric с параметром 'coerce'
data['TotalCharges'] = pd.to_numeric(data['TotalCharges'], errors ='coerce')

In [29]:

data.isnull().any()

customerID          False
BeginDate           False
EndDate             False
Type                False
PaperlessBilling    False
PaymentMethod       False
MonthlyCharges      False
TotalCharges         True
gender              False
SeniorCitizen       False
Partner             False
Dependents          False
InternetService     False
OnlineSecurity      False
OnlineBackup        False
DeviceProtection    False
TechSupport         False
StreamingTV         False
StreamingMovies     False
MultipleLines       False
churn               False
term                False
dtype: bool

In [30]:
# посмотрим пропуски в 'TotalCharges'
data[data['TotalCharges'].isnull()]

customerID  BeginDate    EndDate      Type PaperlessBilling  \
488   4472-LVYGI 2020-02-01 2020-02-01  Two year              Yes   
753   3115-CZMZD 2020-02-01 2020-02-01  Two year               No   
936   5709-LVOEQ 2020-02-01 2020-02-01  Two year               No   
1082  4367-NUYAO 2020-02-01 2020-02-01  Two year               No   
1340  1371-DWPAZ 2020-02-01 2020-02-01  Two year               No   
3331  7644-OMVMY 2020-02-01 2020-02-01  Two year               No   
3826  3213-VVOLG 2020-02-01 2020-02-01  Two year               No   
4380  2520-SGTTA 2020-02-01 2020-02-01  Two year               No   
5218  2923-ARZLG 2020-02-01 2020-02-01  One year              Yes   
6670  4075-WKNIU 2020-02-01 2020-02-01  Two year               No   
6754  2775-SEFEE 2020-02-01 2020-02-01  Two year              Yes   

                  PaymentMethod  MonthlyCharges  TotalCharges  gender  \
488   Bank transfer (automatic)           52.55           NaN  Female   
753                Mailed check           20.25           NaN    Male   
936                Mailed check           80.85           NaN  Female   
1082               Mailed check           25.75           NaN    Male   
1340    Credit card (automatic)           56.05           NaN  Female   
3331               Mailed check           19.85           NaN    Male   
3826               Mailed check           25.35           NaN    Male   
4380               Mailed check           20.00           NaN  Female   
5218               Mailed check           19.70           NaN    Male   
6670               Mailed check           73.35           NaN  Female   
6754  Bank transfer (automatic)           61.90           NaN    Male   

      SeniorCitizen  ... InternetService OnlineSecurity OnlineBackup  \
488               0  ...             DSL            Yes           No   
753               0  ...              No             No           No   
936               0  ...             DSL            Yes          Yes   
1082              0  ...              No             No           No   
1340              0  ...             DSL            Yes          Yes   
3331              0  ...              No             No           No   
3826              0  ...              No             No           No   
4380              0  ...              No             No           No   
5218              0  ...              No             No           No   
6670              0  ...             DSL             No          Yes   
6754              0  ...             DSL            Yes          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies MultipleLines  \
488               Yes         Yes         Yes              No            No   
753                No          No          No              No            No   
936               Yes          No         Yes             Yes            No   
1082               No          No          No              No           Yes   
1340              Yes         Yes         Yes              No            No   
3331               No          No          No              No            No   
3826               No          No          No              No           Yes   
4380               No          No          No              No            No   
5218               No          No          No              No            No   
6670              Yes         Yes         Yes              No           Yes   
6754               No         Yes          No              No           Yes   

     churn term  
488      0    0  
753      0    0  
936      0    0  
1082     0    0  
1340     0    0  
3331     0    0  
3826     0    0  
4380     0    0  
5218     0    0  
6670     0    0  
6754     0    0  

[11 rows x 22 columns]

Пропуски связаны с тем, что клиенты только заключили договор и данных об общих затратах еще нет.

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Хорошо, что обратила внимание на пустые строки в признаке TotalCharges!
    
Согласен, данные пропуски можно заполнить месячными значениями платежей 
    
</div>

In [ ]:
# заполним данные пропуски значением месячных трат
data['TotalCharges'].fillna(data[data['TotalCharges'].isnull()]['MonthlyCharges'],
                                          inplace = True)

In [ ]:
data[data['TotalCharges'].isnull()]

In [ ]:
# поменяем тип данных категориального признака
data['SeniorCitizen'] = data['SeniorCitizen'].astype('object')

Типы данных приведены в порядок.

In [ ]:
data.info()

Рассмотрим столбцы, которые подлежат удалению, в связи с тем, что есть вероятность утечки и последующего переобучения модели на них :
* `BeginDate`,`EndDate` - даты начала и окончания пользования услугами, можно удалить, т.к. на основе этих данных нами были сгенерированы новые признаки;
* `customerID` - уникальный идентификатор лучше тоже исключить.

In [ ]:
data = data.drop(['BeginDate','EndDate','customerID'], axis= 1)

In [ ]:
data.shape

**После проведения предобработки данных можно сделать следующие выводы:**

* 4 исходных датасета объединены в один по общему столбцу `customerID`;
* технологические пропуски после объединения были заменены на 'заглушку' 'not_used' 
    в столбцах `InternetService`, `OnlineSecurity`, `OnlineBackup`,`DeviceProtection`, `TechSupport`, 
    `StreamingTV`, `StreamingMovies`, `MultipleLines`;
* изменены типы данных в столбцах `TotalCharges` и `SeniorCitizen`;   

* создан целевой признак `churn` по условию столбца `EndDate`;
* cоздан дополнительный признак - `term`, срок пользования услугой.
* признаки `BeginDate`, `EndDate`, `customerID` были удалены;
* после изучения всех признаков мы получили датасет размером (7043, 19);

Данные готовы к исследовательскому анализу.

## Исследовательский анализ данных (EDA)


### Изучим целевой признак

Данные содержат численные и категориальные признаки.

Целевой признак  столбец `churn` — факт ухода клиента. Остальные признаки будут участвовать в предсказании решения по столбцу `churn`.
Перед нами стоит задача классификации. Нам требуется спрогнозировать уйдет клиент из телеком компании в ближайшее время или нет.
Посмотрим как распределены классы в целевом признаке `churn`.

In [ ]:
data['churn'].value_counts()

Мы видим существенный дисбаланс классов. Доля клиентов, ушедших из компании составляет около 16%. А доля тех, кто остался 84 %. Наша ключевая метрика оценки качества модели ROC-AUC является не чувствительной к дисбалансу.

### Изучим признаки

Посмотрим матрицу корреляции числовых признаков и выведем тепловую карту для наглядности.

In [ ]:
data.corr()

In [ ]:
plt.figure(figsize=(15, 5))
plt.title('Матрица корреляции численных признаков')

sns.heatmap(data.corr(),annot = True)
plt.show()

Мы видим достаточно слабую корреляцию между числовыми данными и нашим таргетом. Наибольший интерес заслуживает наличие положительной взаимосвязи между сроком пользования услугой и общими затратами клиента (0,82), что является логичным, чем дольше срок тем больше клиент потратил.

Рассмотрим корреляцию смешанного набора признаков, используя phik библиотеку.

In [ ]:
# построим таблицу корреляции смешанных признаков
phik_overview = data.phik_matrix()
# округлим значения коэффициентов для удобства
phik_overview.round(2)

Для наглядности построим тепловую карту матрицы корреляции смешанных признаков.

In [ ]:
plot_correlation_matrix(phik_overview.values, 
                        x_labels=phik_overview.columns, 
                        y_labels=phik_overview.index, 
                        vmin=0, vmax=1, color_map="coolwarm", 
                        title=r'Матрица корреляции',
                        fontsize_factor=1.5, 
                        figsize=(18, 10))
plt.tight_layout();


На тепловой карте мы видим наличие мультиколлинеарности признаков DeviceProtection,TechSupport. 

У нас есть достаточно сильная зависимость наличия дополнительных интернет услуг (особенно InternetService) и ежемесячной стоимости(0.92). Чуть ниже (0.68) зависимость наличия возможности ведения параллельных линий во время звонка (MultipleLines) также на ежемесячную стоимость.

Также мы видим очевидную зависимость между нашим таргетом сроком пользования услугами и опять же общими затратами клиента.

При этом на срок пользования больше всего оказывают влияние такие факторы как общие затраты на услуги в целом (0.85), тип оплаты (0.63) и наличие супруга (0.45).
Пол клиента видимо совершенно не имеет значения.

'Деревянные модели" и бустинги не чувствительны к наличию мультиколлинеарности. А в логистической регрессии реализована регуляризация. Поэтому оставим все как есть.


<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Здорово, что применила библиотеку phik для анализа корреляции смешанного набора признаков!
    
***
    
Что касается корреляции категориальных признаков, то это пока что пища для размышлений, не стоит сразу по этому условию удалять признаки. Данная корреляция может быть нелинейная, и на линейные модели это не повлияет.
    
</div>

### Выведем графики для визуализации наиболее интересных результатов корреляции признаков

Рассмотрим как отличается поведение клиентов в зависимости от разных факторов.

In [ ]:
# выберем стиль для оформления графиков 
plt.style.use('bmh')

In [ ]:
plt.figure()

sns.barplot(x=data['churn'],y=data['Dependents']);
plt.title ('Зависимость от иждивенцев');
plt.xlabel ('Факт ухода');
plt.ylabel ('Наличие иждивенцев');
plt.show()

Ушедших больше среди тех, у кого есть иждивенцы.

In [ ]:
sns.barplot(x=data['churn'],y=data['Partner']);

plt.title ('Зависимость от наличия супруга');
plt.xlabel ('Факт ухода');
plt.ylabel ('Наличие супруга');
plt.show()

Ушедших больше среди тех, у кого есть супруг. Видимо семейные пары чаще пересматривают тарифы на услуги и ищут более выгодные компании, в целях экономии семейного бюджета.

In [ ]:
sns.barplot(x=data['churn'],y=data['gender']);

plt.title ('Зависимость от пола клиента');
plt.xlabel ('Факт ухода');
plt.ylabel ('Пол клиента');
plt.show()

Пол клиента практически не влияет на факт ухода. Видимо этот признак не такой уж существенный.

In [ ]:
sns.histplot(data=data,x='Type',hue='churn', multiple="dodge",shrink=.7)

plt.title('Зависимость типа оплаты и факта ухода')
plt.ylabel('Частота')
plt.xlabel('Тип оплаты')
plt.show();

Больше всего уходят те, кто оплачивает ежемесячно. При длительном договоре вероятность ухода значительно снижается. Поэтому целесообразно рекомендовать компании ежемесячно отслеживать отток клиентов.

In [ ]:
sns.histplot(data=data,x='PaymentMethod',hue='churn', multiple="dodge",shrink=.8)

plt.title('Зависимость способов оплаты и факта ухода')
plt.ylabel('Частота')
plt.xlabel('Способ оплаты')
plt.xticks(rotation = 20)
plt.show();

Ушедших больше среди тех, кому присылают электронный чек, видимо это неудобно для клиента и он испытывает трудности с оплатой.

In [ ]:
sns.histplot(data=data,x='term',hue='churn', multiple="dodge",shrink=.7)

plt.title('Зависимость срока пользования и таргета')
plt.ylabel('Частота')
plt.xlabel('Срок пользования услугами, в днях')
plt.show();

С увеличением срока пользования растет и количество ушедших клиентов.

In [ ]:
sns.histplot(data=data,x='MonthlyCharges',hue='churn', multiple="dodge",shrink=.7)

plt.title('Зависимость ежемесячных платежей и таргета')
plt.ylabel('Частота')
plt.xlabel('Ежемесячные платежи')
plt.show();

При повышении ежемесячных платежей количество ушедших клиентов также возрастает. Особенно это заметно в промежутке ежемесячной стоимости от 70 до 80. Видимо здесь идет критическая отметка и требуется вовремя отслеживать тех клиентов, у которых такие платежи и предлагать им какие-то скидки, чтобы снизить размер платежа.

In [ ]:
data['MonthlyCharges'].hist()

plt.title('Гистограмма значений ежемесячной оплаты'); 
plt.xlabel('Оплата за месяц'); 
plt.ylabel('Частота'); 


Мы видим, что пик суммы ежемесячной оплаты приходится на промежуток от 20 до 30.Видимо это самый минимальный платеж при подключении. Далее идет резкий спад оплаты, возможно это связано с тем, что многие клиенты кто попробовал услугу уходят. Поэтому надо активизировать анализ клиентов, которые подключились недавно и оплатили например только первый месяц.
В промежутке от 70 до 110 остались те клиенты, которых мы удержали и им были подключены дополнительные услуги, поэтому ежемесячный платеж вырос. 

In [ ]:
plt.figure(figsize = (16,8))
data['term'].hist(bins=(50))

plt.title('Гистограмма периода пользования услугами'); 
plt.xlabel('Срок пользования, в днях'); 
plt.ylabel('Частота'); 


In [ ]:
data['term'].describe()

Пик пользования приходится на промежуток от 50 до 200 дней. Далее идет спад и выравнивание. В среднем клиенты пользуются услугами около 2,4 лет.

In [ ]:
sns.histplot(data=data,x='TotalCharges',hue='churn', multiple="dodge",shrink=.7)

plt.title('Зависимость общих затрат и таргета')
plt.ylabel('Частота')
plt.xlabel('Общие затраты')
plt.show();

При повышении общих затрат до 1000 ед. количество ушедших клиентов возрастает. 

In [ ]:
# удалим столбец `gender`
data = data.drop('gender', axis= 1)

In [ ]:
data.shape

**Вывод по EDA**

После проведения исследовательского анализа наших данных можно сделать следующие выводы:

* из исходного датасета было удалено 4 неинформативных столбца: `BeginDate,'EndDate, `customerID`,'gender';
* технологические пропуски в данных были заменены на значение "No";   
* в целевом признаке выявлен дисбаланс классов (доля ушедших 16%);
* дубликаты во всех данных и по customerID не обнаружены;
* после изучения всех признаков мы получили датасет размером (7043, 18);

Данные готовы к обучению моделей.

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Татьяна, хороший анализ данных,  все понятно и очень подробно! Здорово, что по каждому признаку сравнила ушедших и оставшихся клиентов =)
    
*** 
    
Еще нужно только помнить, что влияние различных факторов на целевую переменную, это гипотеза, и нужно использовать методы проверки гипотез. Но это уже отдельная работа =)
    
</div>

## Подготовка данных для обучения моделей

### Разбивка на выборки

Ранее мы ввели константу RANDOM_STATE = 220523

In [ ]:
# cоздадим переменные для признаков и целевого признака

target = data['churn']
features = data.drop('churn', axis=1)

Далее разобьем данные на выборки. Разделим исходные данные на обучающую и тестовую выборки. 
Данные разобьем на 2 части, в соотношении 75:25. Добавим аргумент stratify, он позволит сохранить изначальное распределение таргетов во всех новых датасетах. Существующий дисбаланс никуда не денется, но в каждом датасете он будет одинаковым.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    features, target, test_size=0.25, random_state=RANDOM_STATE,stratify = target
)

In [ ]:
# выведем на экран размеры таблиц, которые хранятся в 4-х переменных 
X_train.shape,y_train.shape,X_test.shape,y_test.shape

## Обучение моделей
### Модель CatBoostClassifier

In [ ]:
# инициализируем модель CatBoostClassifier
#model_cb = CatBoostClassifier(random_state=RANDOM_STATE)

In [ ]:
X_train_cb = X_train.copy()
X_test_cb = X_test.copy()

In [ ]:
num_cols = ['MonthlyCharges', 'TotalCharges', 'DaysUsed']

In [ ]:
# создадим переменную с категориальными признаками для передачи в модель
cat_features = ['Type',
 'PaperlessBilling',
 'PaymentMethod',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'MultipleLines']

In [ ]:
pipeline_catboost = Pipeline([
    ('cb', CatBoostClassifier(cat_features=cat_features,random_state=RANDOM_STATE))
])

parameters_catboost = {
    'cb__iterations': [200,300,400],
    'cb__depth': [3,6,9,12],
    'cb__learning_rate': [0.01, 0.15]
}

In [ ]:
gs_model_cb = GridSearchCV(
    estimator = pipeline_catboost,
    param_grid=parameters_catboost, 
    scoring='roc_auc', 
    cv=KFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE),
    n_jobs=-1)


In [ ]:
gs_model_cb.fit(X_train, y_train)

In [ ]:
# выведем лучшие значения гиперпараметров, полученные на GridSearch

gs_model_cb_best_score = gs_model_cb.best_score_
gs_model_cb_best_params = gs_model_cb.best_params_
print(f'Best params CatBoostClassifier:{gs_model_cb.best_params_}')
print(f'ROC-AUC CatBoostClassifier:{gs_model_cb.best_score_}') 

In [ ]:
# инициализируем лучшую модель с подобранными гиперпараметрами
cb_best = CatBoostClassifier(random_state=RANDOM_STATE,
                            verbose=10,iterations=300,
                            learning_rate=0.15, depth=6,eval_metric='AUC',
                            early_stopping_rounds=20)

In [ ]:
cb_best.fit(X_train_cb, y_train,
                  cat_features = cat_features,verbose=10)

In [ ]:
cb_best.tree_count_

In [ ]:
# построим гистограмму значимости признаков для модели CatBoost
# значения отсортированные по возрастанию влияния

plt.figure(figsize = (16,15))
importances = pd.Series(cb_best.feature_importances_,index=cb_best.feature_names_).sort_values()

# график
importances.plot.barh(grid=True, title='Важность признаков');

Срок пользования услугами оказался самым значимым признаком, что логично. Далее по значимости идут месячные платежи, общие затраты на услуги, тип оплаты и метод оплаты.

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Здорово, что изучила важность признаков 👍
    
Так мы сможем увидеть, какие признаки слабо влияют на отток, и в принципе их можно вообще исключить из обучения, и сделать новую модель, отобрав только самые важные признаки. (Пересчитывать с новыми признаками не нужно).

Анализ важности признаков позволяет не только построить лучшую модель и уменьшить число признаков для обучения. Мы также можем выдать рекомендации бизнесу, изучив влияние признаков на отток, тем самым можно повысить качество определенных услуг, например онлайн-платежи или интернет на оптоволокне.
    
***
    
Вот здесь можно немного почитать про интерпретацию важности признаков
    
https://webiomed.ru/blog/interpretatsiia-rezultatov-mashinnogo-obucheniia/
    
https://habr.com/ru/articles/428213/
    
</div>

### Для кодирования категориальных признаков будем использовать OneHotEncoder и StandardScaler для численных признаков.

In [ ]:
X_train_ridge = X_train.copy()
X_test_ridge = X_test.copy()

In [ ]:
#категориальные признаки для OHE Ridge
ohe_features_ridge = X_train.select_dtypes(include='object').columns.to_list()
print(ohe_features_ridge)


In [ ]:
#численные признаки

num_features = X_train.select_dtypes(exclude='object').columns.to_list()

num_features

In [ ]:

col_transformer_ridge = make_column_transformer(
    (
        OneHotEncoder(drop='first', handle_unknown='ignore',sparse=False),
        ohe_features_ridge
    ),
    (
        StandardScaler(), 
        num_features
    ), remainder='passthrough',
    verbose_feature_names_out=False
)


X_train_ridge = pd.DataFrame(
    col_transformer_ridge.fit_transform(X_train_ridge),
    columns=col_transformer_ridge.get_feature_names_out()
)

#смотрим на результат
X_train_ridge.head()

In [ ]:
# если лучшей моделью будет Ridge, то
# трансформируем тестовую выборку
X_test_ridge = pd.DataFrame(
    col_transformer_ridge.transform(X_test_ridge),
    columns=col_transformer_ridge.get_feature_names_out()
)

#смотрим на результат
X_test_ridge.head()

In [ ]:
#получились одинаковые признаки в тренировочной и тестовой выборках
(X_train_ridge.columns != X_test_ridge.columns).sum()

In [ ]:
X_train_ridge = X_train.copy()

### Модель Ridge

In [ ]:
#random_state не перебирается, задаём его прямо в модели
model_ridge = Ridge(random_state=RANDOM_STATE)

pipeline_ridge = make_pipeline(col_transformer_ridge, model_ridge)

param_grid_ridge = {
    'ridge__alpha': np.arange(0, 0.21, 0.01),
}

gs_ridge_pl = GridSearchCV(
    pipeline_ridge, 
    param_grid=param_grid_ridge, 
    scoring='roc_auc',
    n_jobs=-1,error_score='raise'
)


In [ ]:
gs_ridge_pl.fit(X_train_ridge, y_train)

In [ ]:
# лучшее значение ROC-AUC на кросс-валидации
print(f'ROC-AUC Ridge: {gs_ridge_pl.best_score_}')
# лучшие гиперпараметры
print(f'best_params Ridge: {gs_ridge_pl.best_params_}')

In [ ]:
#features_train = X_train.copy()
#features_test = X_test.copy()

In [ ]:
# категориальные признаки для OHE
#ohe_features = features_train.select_dtypes(include='object').columns.to_list()
#ohe_features

In [ ]:
# drop='first' удаляет первый признак из закодированных:
# таким образом обходим dummy-ловушку
# задаём handle_unknown='ignore':
# игнорируется ранее не встречающиеся значения признака (при transform)
#encoder_ohe = OneHotEncoder(drop='first', handle_unknown='ignore',sparse=False)

# обучаем энкодер на заданных категориальных признаках тренировочной выборки
#encoder_ohe.fit(features_train[ohe_features])

# добавляем закодированные признаки в features_train
# encoder_ohe.get_feature_names_out() позволяет получить названия колонок
#features_train[encoder_ohe.get_feature_names_out()] = encoder_ohe.transform(features_train[ohe_features])

# удаляем незакодированные категориальные признаки (изначальные колонки)
#features_train_rf = features_train.drop(ohe_features, axis=1)

# смотрим на результат
#features_train_rf.head()

In [ ]:
# энкодером, который обучен на тренировочной, кодируем тестовую в случае если будет достигнута метрика качества
#features_test[encoder_ohe.get_feature_names_out()] = encoder_ohe.transform(features_test[ohe_features])
#features_test_rf = features_test.drop(ohe_features, axis=1)

Далее в целях подготовки данных для дальнейшего исследования проведем масштабирование численных признаков методом стандартизации, чтобы показать модели, что все признаки значимы.

In [ ]:
#pd.options.mode.chained_assignment = None

# список численных признаков для стандартизации
#numeric = ['MonthlyCharges', 'TotalCharges', 'term']

# создадим объект структуры для стандартизации данных — StandardScaler
#scaler = StandardScaler()
# настроим объект на обучающих данных
#scaler.fit(features_train[numeric]) 

# преобразуем все выборки функцией transform
#features_train[numeric]= scaler.transform(features_train[numeric])

#features_test[numeric] = scaler.transform(features_test[numeric])

# проверим размеры выборок и выведем на экран стандартизированные столбцы обучающей выборки
#print(features_train[numeric].shape,features_test[numeric].shape)


In [ ]:
#features_train[numeric]

In [ ]:
# удаляем незакодированные численные признаки (изначальные колонки)
#features_train_rf = features_train_rf.drop(features_train[numeric], axis=1)
# объединяем 2 датафрейма
#features_train_rf = pd.concat([features_train_rf,features_train[numeric]], axis=1)

#features_train_rf

In [ ]:
# тоже самое для тестовой выборки
#features_test_rf = features_test_rf.drop(features_test[numeric], axis=1)
#features_test_rf = pd.concat([features_test_rf,features_test[numeric]], axis=1)
#features_test_rf.shape

In [ ]:
# проверим размеры
#features_train_rf.shape

### Модель случайного леса классификации RandomForestClassifier

In [ ]:
#категориальные признаки для OHE RandomForestClassifier
ohe_features_rf = ohe_features_ridge.copy()
ohe_features_rf

In [ ]:
#численные признаки
num_features = X_train.select_dtypes(exclude='object').columns.to_list()
num_features

In [ ]:
X_train_rf = X_train.copy()
X_test_rf = X_test.copy()

In [ ]:


col_transformer_rf= make_column_transformer(
    
    (
        OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1),   
        ohe_features_rf
    ),
    
    (StandardScaler(), 
        num_features
    )
)


In [ ]:
model_rf = RandomForestClassifier(random_state=RANDOM_STATE)

pipline_rf = make_pipeline(col_transformer_rf, model_rf)

# подберем гиперпараметры, используя  GridSearchCV
param_rf = { 
    'n_estimators': [100, 150, 200],
    'max_depth' : [2, 5, 10, 20, 25, 30],
    'class_weight': ['balanced']
}

gs_model_rf = GridSearchCV(
    estimator=model_rf, 
    param_grid=param_rf, 
    scoring='roc_auc',error_score='raise',
     n_jobs=-1)



gs_model_rf.fit(X_train_rf, y_train)



In [ ]:
# лучшее значение ROC-AUC на кросс-валидации
print(f'best_score RandomForestClassifier: {gs_model_rfc.best_score_}')
# лучшие гиперпараметры
print(f'best_params RandomForestClassifier: {gs_model_rfc.best_params_}')      

In [ ]:
model_rfc = RandomForestClassifier(random_state=RANDOM_STATE)

pipline_rfc = make_pipeline(model_rfc)

# подберем гиперпараметры, используя  GridSearchCV
param_rfc = { 
    'n_estimators': [100, 150, 200],
    'max_depth' : [2, 5, 10, 20, 25, 30],
    'class_weight': ['balanced']
}

gs_model_rfc = GridSearchCV(
    estimator=model_rfc, 
    param_grid=param_rfc, 
    scoring='roc_auc',
    cv=5, n_jobs=-1)



gs_model_rfc.fit(features_train_rf, y_train)



In [ ]:

# лучшее значение ROC-AUC на кросс-валидации
print(f'best_score RandomForestClassifier: {gs_model_rfc.best_score_}')
# лучшие гиперпараметры
print(f'best_params RandomForestClassifier: {gs_model_rfc.best_params_}')      

In [ ]:
pd.DataFrame({
        'Model':['Ridge','RandomForestClassifier','CatBoostClassifierr'],
        'ROC-AUC':['0.76','0.81','0.89']
})

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Здорово, что рассмотрела несколько моделей и сделала перебор гиперпараметров! 
     
Видно, что проделана большая работа 👍 Обучение моделей выполнено правильно
    
</div>

### Тестирование лучшей модели на тестовой выборке

In [ ]:
# инициализируем модель случайного леса с наилучшим значением метрики ROC-AUC
best_model = CatBoostClassifier(
                            random_state=RANDOM_STATE,
                            verbose=10,iterations=300,
                            learning_rate=0.15, depth=6,eval_metric='AUC',
                            early_stopping_rounds=20
)

best_model.fit(X_train_cb, y_train,cat_features = cat_features) 
predict_test = best_model.predict(X_test_cb)

### Построение ROC кривой и расчет дополнительных метрик качества

Построим ROC-кривую для данной модели. Для этого найдем значения вероятностей классов для тестовой выборки. Значения вероятностей класса «1» сохраним в переменной probabilities_one_test.

In [ ]:
probabilities_test = best_model.predict_proba(X_test)
probabilities_one_test = probabilities_test[:, 1]

fpr, tpr, thresholds = roc_curve(y_test, probabilities_one_test)
plt.figure()
plt.plot(fpr, tpr)
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.0])
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC-кривая')
plt.show() 

In [ ]:
# выведем матрицу ошибок
cm = confusion_matrix(y_test, predict_test, labels=best_model.classes_)
disp = ConfusionMatrixDisplay(confusion_matrix=cm,
                           display_labels=['Non-churned', 'Churned'])
disp.plot()
plt.show()

In [ ]:
# рассчитаем дополнительные метрики
print(f'Accuracy: {accuracy_score(y_test, predict_test)}')    
print(f'Полнота: {recall_score(y_test, predict_test)}')     
print(f'Точность: {precision_score(y_test, predict_test)}')  
print(f'F1-мера: {f1_score(y_test,predict_test)}')          
print(f'AUC-ROC:{roc_auc_score(y_test,probabilities_one_test)}')


<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Отличные результаты на тестировании, поздравляю!
    
</div>

Accuracy мы получили 0.90, но в нашем случае при дисбалансе классов данная метрика не совсем информативна.

Recall мы получили лишь 0.53, это говорит нам о том, что модель из всех склонных к оттоку клиентов находит лишь 53%.
Но при этом мы получили высокое значение Precision 0.82. Это говорит о том, что наша модель находит 82% склонных к оттоку и при этом они действительно уходят.

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Вот небольшие статьи по метрикам в машинном обучении, если заинтересует:

https://habr.com/ru/company/ods/blog/328372/ 
    
https://habr.com/ru/company/jetinfosystems/blog/420261/

</div>

<div class="alert alert-info"> <b>ℹ️ Комментарий студента: </b> Спасибо за статьи. Скажи пожалуйста, по метрикам до конца не уложится в голове: вот в нашем случае нам нужно, чтобы recall был все-таки выше, правильно я понимаю? чтобы лучше находить из всех клиентов именно склонных к оттоку, бизнесу важнее же это? и как нам повысить эту метрику если это нужно? с другой стороны мы же достигли требуемого AUC-ROC... </div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера v2:</b> По условию задачи мы максимизируем метрику ROC-AUC, но с точки зрения бизнеса нужны немного другие результаты. Ты задала хороший вопрос =)
    
Метрика ROC-AUC не зависит от порога классификаци, но метрику Recall можно улучшить, изменяя порог классификации (по умолчанию 0,5).

По поводу метрики для бизнеса давай разберем пример 😊

1. Сначала рассмотрим пример, касающийся медицины или диагностики заболеваний:     

- в этом случае доля ложно-отрицательных ответов должна быть сведена к нулю. Если, к примеру вероятность заболевания > 5% (p = 0.05), то мы относим пациента в группу риска. Мы конечно сильно ухудшим метрику F1,  но целевой метрикой в данном случае является Recall. Даже если пациент здоров, он пройдет дополнительное обследование    

2. В задачах классификации с точки зрения бизнеса значение метрики определяется требованиями бизнеса, можно даже поставить задачу — чем больше, тем лучше. Но хороший показатель ROC-AUC не всегда означает хорошую модель, важно правильно выбрать порог классификации.    

Представим, что у нас промокод, который составляет 30% от стоимости ежемесячных услуг для каждого клиента. Поставим задачу так: кто собирается уходить, тому выдадим промокод, а кто лояльный, промокод не будем давать. У нас появятся ошибки первого и второго рода.
    
Ошибки первого рода: 
- предсказание модели 1, значение таргета 0, мы выдали промокод лояльному клиенту.
    
Ошибки второго рода:
- предсказание модели 0, значение таргета 1, не выдали промокод клиенту, который склонен к оттоку.    

Чтобы посчитать прибыль, рассмотрим такую схему:    

- если клиент на самом деле был лояльный, то мы потеряли 30% от стоимости месячного платежа (ошибка 1го рода)
- если клиент был отточный, а мы не дали ему промокод, то мы потеряем 100% от стоимости ежемесячного платежа (ошибка второго рода).
    
По умолчанию порог классификации равен 0,5, можно сделать разные расчеты, кода порог равен 0.5, 0.6, 0.7, 0.8, 0.9. 
На тестовой выборке рассчитваем убытки по ошибкам первого и второго рода, соответственно наша задача — минимизация убытков.
    

</div>

<div class="alert alert-info"> <b>ℹ️ Комментарий студента 2: </b> Супер, спасибо за примеры, так гораздо яснее. Скажи пожалуйста а где мы этот порог меняем, в каких параметрах? Либо у нас этого не было, либо забыла... И что значит на тестовой выборке мы рассчитываем убытки, как это делается🤔?</div>

<div class="alert alert-success">
<b>✔️ Комментарий тимлида v2:</b> Нет, такую задачу мы не рассматривали, я ее иногда предлагаю рассмотреть, кому интересно, как дополнение к проекту =)
    
Напишу об этом завтра в Пачке, может кто-то заинтересуется
    
    
Задача в том, кому выдать промокоды, чтобы минимизировать убытки
    
</div>

Проверим на адекватность.

In [ ]:
dummy_model = DummyClassifier(strategy='constant', constant=1)
dummy_model.fit(features_train,target_train)
dummy_model.score(features_train,target_train)

Получили, что константная модель в 16 % случаев будет предсказывать класс 1, что соответсвует распределению классов в таргете.

Лучшей моделью, которая достигла требуемой метрики качества ROC-AUC>=0.85 вышла модель **CatBoostClassifier** с подобранными гиперпараметрами random_state=RANDOM_STATE, iterations=300,learning_rate=0.15, depth=6. AUC-ROC на тестовых данных:0.9064272604918635. Точность: 0.8114285714285714.

Более подробный вывод напишу в отчетной части.

<font color=blue>
<b> ✔️ 👍Заключение ревьювера:</b> Татьяна, было приятно проверять твою работу, она выполнена на очень хорошем уровне! 

Тебе удалось справиться с выпускным проектом, поздравляю! Данные из разных таблиц собраны в один датафрейм, удалены аномалии, заполнены пропуски, получены новые признаки, а ненужные признаки удалены. Рассмотрено несколько моделей и сделан перебор параметров. Могу отметить самостоятельность и аналитический подход в выполнении проекта, хорошую аналитическую часть!
   
Положительные моменты: 
    
- грамотное и понятное оформление кода;
    
- аналитический подход — при удалении признаков и формировании новых написано соответствующее обоснование;
    
- проведен исследовательский анализ данных и корреляции признаков;
    
- рассмотрены несколько моделей для обучения, выполнен перебор гиперпараметров.

Особых замечаний по работе не возникло.    
    
На данном этапе проект практически завершен, я также отметил рекомендации. У нас еще есть время — я отправлю проект, чтобы ты могла ознакомиться с проверкой и возможно задать вопросы =)
    
Также на будущее стоит иметь ввиду, что метрика ROC-AUC является не единственной, всегда нужно прислушиваться к требованиями бизнеса. Обычно одной метрикой не ограничиваются, смотрят также Precision, Recall, и выбирают те, которые наиболее подходят для конкретных задач. 
    
Еще нужно смотреть на важность признаков в модели, чтобы понимать, какие параметры влияют на отток.
    
Проверка важности признаков по модели это только один из способов (корреляция в том числе). Используют также аналитические методы проверки гипотез.
    
Если есть какие-то вопросы, или нужны пояснения по проекту — смело пиши! Я постараюсь тебе помочь 😊

Жду твоего ответа!
    
</font>

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера v2:</b> Татьяна, поздравляю с выполнением основной части проекта!
    
Пожалуйста, сохрани наши с тобой комментарии, а отчет продолжай выполнять ниже, в этом файле Jupyter Notebook.
    
Вижу, что ты уже практически сделала отчет, на всякий случай напомню необходимые пункты:
       
- список признаков, которые использовались для обучения;
    
- указать итоговую модель и ее гиперпараметры — желательно в отдельной строке, чтобы не сливалась с остальным текстом;
    
- отличия от исходного плана работ;
    
- в отчете нужно указать трудности в работе (если не было, то так и написать) и ключевые этапы (если ключевых этапов не было, то написать, что все этапы важны);
    
- размер тестовой выборки, параметр random_state.

Сейчас необходимо добавить список признаков, которые использовались для обучения моделей.
    
</div>

<div class="alert alert-info"> <b>ℹ️ Комментарий студента 2: </b> Признаки добавила.</div>

<div class="alert alert-success">
<b>✔️ Комментарий тимлида v2:</b> Отлично
    
</div>

## Общий вывод 

**Вывод**   

Мы провели работу по подготовке прототипа модели машинного обучения для оператора связи «Ниединогоразрыва.ком» 

**Цель проекта** - прогнозирование оттока клиентов.   

**Наша задача** - создание прототипа модели машинного обучения, который позволит оператору связи своевременно выявлять клиентов, планирующих уйти в ближайшее время. Это позволит компании предложить клиентам промокоды и специальные условия, тем самым сохранить клиента и увеличить прибыль.  

В нашем распоряжении были персональные данные о некоторых клиентах,информация об их тарифах и договорах, которые собрала команда оператора.

Данные находились в 4-х датасетах. В ходе работы все данные были объединены в один датасет по общему признаку `CustomerID`.

Работа была проведена в соответствии с первононачальным планом. Сложностей в процессе работы не возникло.

Была проведена загрузка исходных данных, их обзор и предобработка: 

* в столбце `EndDate` не везде была отмечена дата окончания пользования услугами. В ходе анализа мы заменили данные пропуски на дату выгрузки - 01.02.2020.
* найденные технологические пропуски после объединения датасетов были заполнены значениями "not_used". 
* дубликаты в данных отсутствовали. 
* аномалий не обнаружено. 
* создали столбец с таргетом `churn` на основе данных из столбца `EndDate` -  где '1' - клиент ушел, а '0' - клиент действующий.
* cоздали дополнительный признак - `term`, срок пользования услугой в днях;
* неинформативные (`CustomerID`) и столбцы, приводящие к утечке данных (`BeginDate`,`EndDate`)были удалены после создания таргета и дополнительного признака, т.к. не представляют важности при обучении модели. 
* привели в соответствие типы данных в столбцах `TotalCharges` и `SeniorCitizen`;
* обнаруженные пропуски в столбце `TotalCharges` были заменены на значение месячных платежей;

Далее мы провели исследовательский анализ данных и визуализировали с помощью графиков результаты зависимости поведения клиентов в двух группах от различных факторов и дополнительных услуг:
* изучили целевой признак и обнаружили дисбаланс классов, класса с меткой '1'было около 16%,оставшиеся были с меткой '0'.
* построили тепловую карту корреляции числовых признаков, выявили наличие положительной взаимосвязи между сроком пользования услугой и общими затратами клиента (0,82), что является логичным, чем дольше срок тем больше клиент потратил
* посмотрели на графиках наиболее интересные зависимости поведения клиентов от различных признаков;
* выяснили, что пол клиента практически не имеет значения, поэтому впоследствии данный признак был удален;
* на срок пользования больше всего оказывают влияние такие факторы как общие затраты на услуги в целом, тип и способ оплаты, наличие супруга.
* после изучения всех признаков мы получили датасет размером (7043, 18), готовый для обучения моделей.

Следующим шагом была проведена разбивка данных на обучающий и тестовый набор, с **размером тестовой выборки 0.25**. При разбивке добавили аргумент stratify, который позволил сохранить изначальное распределение таргета во всех новых датасетах. Существующий дисбаланс никуда не денется, но в каждом датасете он будет одинаковым.
Был введен параметр  **RANDOM_STATE = 220523**.
Были выделены признаки и целевой признак. 

**Так для обучения модели были использованы следующие 17 признаков:**

Type,PaperlessBilling,	PaymentMethod,	MonthlyCharges	TotalCharges,   
SeniorCitizen,Partner,Dependents,InternetService,OnlineSecurity,OnlineBackup,   	DeviceProtection,TechSupport,StreamingTV,StreamingMovies,MultipleLines,term.   

Далее мы провели обучение моделей машинного обучения, для выполнения поставленной цели. 
Ключевой метрикой качества был ROC-AUC, требовалось достичь значения >=0.85.

Нами было инициализировано 3 модели: Ridge, RandomForestClassifier, CatBoostClassifier.
Для подбора гиперпараметров в целях улучшения качества моделей, а также для проведения проверки качества был использован GridSearchCV.

Для моделей Ridge и RandomForestClassifier было проведено кодирование категориальных признаков методом OneHotEncoder и масштабирование численных признаков методом стандартизации StandardScaler.

Выбрав лучшую модель, которая достигла требуемого значения ROC-AUC,мы рассчитали ее качество на тестовой выборке. Также была построена ROC-кривая, рассчитаны дополнительные метрики качества и проведена проверка на адекватность лучшей модели в сравнении с константной моделью.

Также мы рассмотрели наиболее значимые признаки, которые позволили достигнуть нашей метрики.
Можем выделить следующие **наиболее важные признаки** при обучении модели:
* term;
* TotalCharges;
* Type;
* MonthlyCharges;
* PaymentMethod;
* MultipleLines.

В этой связи рекомендуем бизнесу провести более детальный анализ значимых признаков и провести работу по улучшению качества данных услуг в целях максимизации прибыли.
		
Итак, лучшей моделью машинного обучения в нашем случае выбрана модель CatBoostClassifier, показав на тестовых данных значение 
ROC-AUC 0.90 (на обучающей 0.89).

**Гиперпараметры итоговой лучшей модели     
CatBoostClassifier(random_state=RANDOM_STATE,       
                   iterations=300,      
                   learning_rate=0.15,       
                   depth=6,verbose=0)**   

Исходя из критериев которые важны заказчику, **рекомендуем компании выбрать модель  CatBoostClassifier**. 

По матрице ошибок мы можем сказать, что количество истинно положительных ответов (то есть тех кто по прогнозу модели ушли из компании и фактически ушли) достаточно большое, что позволит нам использовать модель в своих прогнозах, не расходуя рекламный бюджет на привлечение новых клиентов. 

Недостатком данной модели является недостаточно высокое значения метрики recall - 0.51.Для бизнеса важно максимально точно отделить тех, кто хочет уйти от тех, кто бы остался. Если клиент хочет уйти в отток, то следует предложить ему скидку или выгодный тариф, чтобы он остался. А если он итак не собирался уходить, то не стоит тратить средства на его удержание. Поэтому следует еще провести работу над повышением данной метрики, возможно удалением части незначимых признаков.

Для бизнеса это важно, потому что модель достаточно хорошо находит тех клиентов, которые собираются уйти, а это значит, что компания может принять меры для удержания клиента и получения прибыли от него в дальнейшем длительном периоде.

Метрика Precision говорит о том, что наша модель правильно определяет на 81% из всех клиентов именно клиентов склонных к оттоку и при этом они действительно уходят.

Для моделей Ridge и RandomForestClassifier не удалось достичь требуемого значения ROC-AUC.
Техника борьбы с дисбалансом (установка параметра class_weight = 'balanced') в данном случае также не помогли улучшить модели. Соответственно данные модели нам не подходят.

Дана рекомендация бизнесу, активно работать с теми клиентами,чье поведение (по прогнозам нашей модели) свидетельствует о том, что они уйдут из банка в ближайшее время.

<div class="alert alert-info"> <b>ℹ️ Комментарий студента: </b> Арсен, спасибо за рекомендации. Еще поясни пожалуйста, что нам дает применение  class_weight = 'balanced', я пробовала его на Catboost, но с ним метрика чуть хуже. Почему все рекомендуют его применять при дисбалансе, а он ничего особо не дает? И еще вопрос: на практике такие модели по определению оттока строятся на сотнях признаков или это может быть 30 признаков, но крутых? и сколько по времени может занимать процесс отбора признаков и подготовки данных? какие задачи у джунов на практике, что им в основном поручают делать? Спасибо </div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера v2:</b> Татьяна, тебе спасибо за хорошую работу!
    
Метрика ROC-AUC не чувствительна к дисбалансу, но для других метрик дисбаланс имеет смысл учитывать.
    
***
    
На практике бывают по-разному: 
    
- сотни разных признаков и несколько десятков значимых признаков могут дать хорошие результаты. По возможности конечно шумные признаки лучше убирать.
    
Генерация признаков — это одна из важных задач на этапе подготовки данных, эта задача может занимать несколько дней, в зависимости от сложности =) В соревнованиях на Kaggle участники постоянно улучшают модели, генерируют новые признаки на протяжении всего соревнования, которое может длиться месяц.
    
Отбор признаков можно сделать на основе проверки гипотез о значимости признака:
    
- для общего развития можно посмотреть цикл статей:

https://habr.com/ru/post/557424/
https://habr.com/ru/post/556798/
</div>

<div class="alert alert-info"> <b>ℹ️ Комментарий студента 2: </b> Отчет дописала. Про дисбаланс поняла. Ты еще обещал рассказать про задачи джунов). Также интересно было бы послушать про твой опыт, чем занимаешься сейчас, сколько в DS,IT😊 Спасибо еще раз за развернутые ответы своими словами понятным языком👍</div>

<div class="alert alert-success">
<b>✔️ Комментарий тимлида v2:</b> Все хорошо, отчет принимается =)
    
По поводу джунов: 
    
- часто джуны решают задачи выгрузки данных из базы (SQL), проверка гипотез, расчет метрик, визуализация данных, генерация признаков, ревью кода коллег, исправления кода под новые версии библиотек и многие другие задачи. Это я описал предметную области машинного обучения и аналитики.
    
Более подробно я обычно рассказываю на финальном вебинаре, так что приходи =)
    
В DS я с 2020 года, в свое время тоже прошел курс по Машинному обучению, в прошлом работал инженером-конструктором =) Сейчас работаю в области ML-аналитики. Подробнее также могу рассказать на вебинаре 😉
    
</div>

<div class="alert alert-success">
<b>✔️ Заключение тимлида:</b> Татьяна, было приятно поработать с тобой над финальным проектом 😊 Видно, что ты серьезно отнеслась к работе, как к реальной задаче!
        
✨ Поздравляю тебя с успешным завершением финального спринта! Если вспомнить, с чего начиналось обучение, видно как от простых задач мы перешли к реальным проектам, и у тебя это здорово получилось. Я считаю, что это заслуживает похвалы 👍
    
В процессе работы над проектом ты проявила себя как грамотный специалист — могу отметить самостоятельность и аналитический подход! По отчету понятно, какие признаки использовались для обучения, как были получены дополнительные признаки, указана модель для обучения и гиперпараметры — по этим шагам можно воспроизвести проект.
    
Сейчас ты можешь доработать этот проект с учетом рекомендаций, удалить наши с тобой комментарии и добавить себе в портфолио — лучше это сделать в ближайшее время, пока все нюансы еще отложились в памяти =) Как рекомендация, можешь попробовать Pipeline для подготовки признаков и обучения моделей
     
Здесь хочу добавить несколько рекомендаций на будущее:
    
- для анализа данных по ситуации можно применять библиотеки pandas_profiling или sweetviz. Когда нужен оперативный отчет по датафрейму, можно воспользоваться такими инструментами для автоматизации. Но в учебных проектах мы делали вручную, поскольку должны были всему научиться;
    
- следует отметить, что можно использовать аналитические методы проверки гипотез для оценки влияния признаков на отток. Например, можно воспользоваться инструментом ANOVA для дисперсионного анализа: https://www.reneshbedre.com/blog/anova.html
    
***

Далее, хочу добавить некоторые ссылки на полезные ресурсы по машинному обучению:

https://academy.yandex.ru/handbook/ml онлайн-учебник от Школы анализа данных Яндекса, в котором описаны теоретические основы работы моделей машинного обучения;

https://www.youtube.com/watch?v=xl1fwCza9C8 познавательное видео по настройке модели CatBoost

https://habr.com/ru/company/ods/blog/322626/ на Habr можно закрепить свои знания, порешав задачи из цикла статей — Открытый курс машинного обучения

https://github.com/esokolov/ml-course-hse — на гитхаб есть репозиторий с задачами из курса по машинному обучению от Евгения Соколова, можно использовать как дополнительный материал для закрепления знаний.

https://habr.com/ru/company/avito/blog/571094/ — материалы по A/B тестам.

Также хочу поделиться опытом и вкратце рассказать о том, в какие направления можно подаваться на работу и какие навыки там пригодятся:

Направление аналитики: хорошие знания теории вероятностей и математической статистики; базовые знания библиотек ML; уверенные знания SQL; умение решать задачи по A/B тестам; плюсом будет знание специальных инструментов для аналитики по визуализации результатов.

Направление ML: уверенные знания классических моделей машинного обучения; хорошие знания SQL; понимание алгоритмов машинного обучения — как устроена модель линейной регрессии, модели случайного леса, градиентного бустинга и т.д. На собеседовании могут спросить, как работает какая-нибудь модель «под капотом», очень любят градиентный бустинг и случайный лес. Здесь поможет учебник https://academy.yandex.ru/handbook/ml

В некоторых компаниях при устройстве на работу или стажировку, например в Яндексе, нужно решить тест на алгоритмические задачи. По алгоритмам есть много разных курсов, платных и бесплатных. Можно попробовать Тренировки по Алгоритмам от Яндекса https://yandex.ru/yaintern/algorithm-training (бесплатно, хороший курс). Насколько знаю, у Практикума тоже есть курс по алгоритмам.

Есть компании, где требуется умение решать задачи с использованием нейронных сетей, в основном задачи распознавания изображений и анализа текстов (NLP). Если понравилось работать с нейронками, желательно порешать разные задачи, набрать портфолио из нескольких работ, можно подаваться.

Существуют научные и производственные компании, которые внедряют алгоритмы ML в свои рабочие процессы — в таких компаниях можно совместить должность, например инженера и программиста. Тоже хороший вариант, но нужно понимать, если хотите развиваться дальше, особенно на первых этапах, у вас должны быть опытные наставники, которые помогут советом.
    
***
    
После окончания финального спринта у нас будет заключительная консультация, где мы сможем подвести итоги и обсудить несколько проектов. Есть возможность выступить перед сокурсниками со своим проектом в zoom =)
    
Отчет принимаю и до встречи!
    
</div>